src/core/server/types.ts

```diff
@@ -17,12 +17,4 @@
  * under the License.
  */
 
-import { SavedObjectsClient } from '../../saved_objects';
-
-export function initSavedObjectClient(chrome) {
-  const savedObjectClient = new SavedObjectsClient();
-
-  chrome.getSavedObjectsClient = function () {
-    return savedObjectClient;
-  };
-}
+export * from './saved_objects/types';
```

Comment:
I should have written it down, because after my leave I don't remember the exact error message, but for some reason consuming the saved object types from `/public` caused problems with our build system as some server-side code ended up being included in the bundle. I'll investigate it again.

Reply:
When I import `SavedObjectsMigrationVersion` from `src/core/server` I get the following:

```
   │ proc [kibana]  FATAL  Error: Optimizations failure.
     │ proc [kibana]    2847 modules
     │ proc [kibana]     
     │ proc [kibana]     WARNING in ./src/core/server/plugins/plugin.ts 215:29-54
     │ proc [kibana]     Critical dependency: the request of a dependency is an expression
     │ proc [kibana]     
     │ proc [kibana]     WARNING in ./src/core/server/plugins/plugin.ts 233:29-74
     │ proc [kibana]     Critical dependency: the request of a dependency is an expression
     │ proc [kibana]     
     │ proc [kibana]     WARNING in ./src/legacy/plugin_discovery/plugin_pack/create_pack.js 33:17-51
     │ proc [kibana]     Critical dependency: the request of a dependency is an expression
     │ proc [kibana]     
     │ proc [kibana]     WARNING in ./node_modules/thread-loader/dist/worker.js 258:32-42
     │ proc [kibana]     Critical dependency: the request of a dependency is an expression
     │ proc [kibana]     
     │ proc [kibana]     ERROR in ./node_modules/thread-loader/dist/WorkerPool.js
     │ proc [kibana]     Module not found: Error: Can't resolve 'child_process' in '/Users/rudolf/dev/kibana/node_modules/thread-loader/dist'
     │ proc [kibana]     
     │ proc [kibana]     ERROR in ./src/cli/cluster/worker.js
     │ proc [kibana]     Module not found: Error: Can't resolve 'cluster' in '/Users/rudolf/dev/kibana/src/cli/cluster'
     │ proc [kibana]     
     │ proc [kibana]     ERROR in ./src/core/server/bootstrap.ts
     │ proc [kibana]     Module not found: Error: Can't resolve 'cluster' in '/Users/rudolf/dev/kibana/src/core/server'
     │ proc [kibana]     
     │ proc [kibana]     ERROR in ./src/legacy/server/kbn_server.js
     │ proc [kibana]     Module not found: Error: Can't resolve 'cluster' in '/Users/rudolf/dev/kibana/src/legacy/server'
     │ proc [kibana]     
     │ proc [kibana]     ERROR in ./src/optimize/watch/watch.js
     │ proc [kibana]     Module not found: Error: Can't resolve 'cluster' in '/Users/rudolf/dev/kibana/src/optimize/watch'
     │ proc [kibana]     
     │ proc [kibana]     ERROR in ./node_modules/thread-loader/dist/worker.js
     │ proc [kibana]     Module not found: Error: Can't resolve 'module' in '/Users/rudolf/dev/kibana/node_modules/thread-loader/dist'
     │ proc [kibana]     
     │ proc [kibana]     ERROR in ./src/cli/cluster/cluster_manager.js
     │ proc [kibana]     Module not found: Error: Can't resolve 'readline' in '/Users/rudolf/dev/kibana/src/cli/cluster'
     │ proc [kibana]     
     │ proc [kibana]     ERROR in ./src/legacy/server/sample_data/routes/lib/load_data.js
     │ proc [kibana]     Module not found: Error: Can't resolve 'readline' in '/Users/rudolf/dev/kibana/src/legacy/server/sample_data/routes/lib'
     │ proc [kibana]     
     │ proc [kibana]     ERROR in ./src/cli/repl/index.js
     │ proc [kibana]     Module not found: Error: Can't resolve 'repl' in '/Users/rudolf/dev/kibana/src/cli/repl'
     │ proc [kibana]     
     │ proc [kibana]     ERROR in ./src/legacy/server/status/lib/metrics.js
     │ proc [kibana]     Module not found: Error: Can't resolve 'v8' in '/Users/rudolf/dev/kibana/src/legacy/server/status/lib'
```

Reply:
It seems that we fail on importing server-specific modules. I'm wondering if tree shaking should handle this case 🤔 or we have to reorganize type exporting completely

Reply:
Could the problem be caused by these changes?
https://github.com/elastic/kibana/pull/39891/files#diff-f1fc96b9d33d05eec20823c52a2ba7ffL24

Reply:
I ran into this issue when building the client-side PluginsService. I don't remember exactly which incantation of exports/imports fixed it, but it had something to do with avoiding any `export * from` declarations in the server exports. I'll see if I can't get this to work on this branch.

Reply:
I had something similar when I was I was working on sharing some config between `public` & `server` in the Integrations Manager plugin. I'll try to find the failing CI build to show, but iirc, I had something in `common` that `import`ed from `server` (e.g. reached "down"), so when `public` imported from `common` it pulled in the `server` deps.  It seems obvious now, but I had the stuff in server because I thought it was largely the server's domain, even if others might use it. Solution was to lift stuff "up" to common and have `public` and `server` pull from `common`.

More concretely, perhaps a something like [`dependency-cruiser`](https://github.com/sverweij/dependency-cruiser) would help find the issue.

e.g. you could do `depcruise --exclude "^node_modules" --output-type dot src/core/server/saved_objects/migrations/core/document_migrator.ts` then paste the output into [an online dot viewer](https://dreampuf.github.io/GraphvizOnline/#strict%20digraph%20%22dependency-cruiser%20output%22%7B%0A%20%20%20%20ordering%3Dout%0A%20%20%20%20rankdir%3DLR%0A%20%20%20%20splines%3Dtrue%0A%20%20%20%20overlap%3Dfalse%0A%20%20%20%20nodesep%3D0.16%0A%20%20%20%20ranksep%3D0.18%0A%20%20%20%20fontname%3D%22Helvetica-bold%22%0A%20%20%20%20fontsize%3D9%0A%20%20%20%20style%3D%22rounded%2Cbold%2Cfilled%22%0A%20%20%20%20fillcolor%3D%22%23ffffff%22%0A%20%20%20%20compound%3Dtrue%0A%20%20%20%20node%20%5Bshape%3Dbox%20style%3D%22rounded%2C%20filled%22%20fillcolor%3D%22%23ffffcc%22%20height%3D0.2%20fontname%3DHelvetica%20fontsize%3D9%5D%0A%20%20%20%20edge%20%5Bcolor%3D%22%2300000077%22%20penwidth%3D2.0%20arrowhead%3Dnormal%20fontname%3DHelvetica%20fontsize%3D9%5D%0A%0A%20%20%20%20subgraph%20%22cluster_..%22%20%7Blabel%3D%22..%22%20subgraph%20%22cluster_..%2F..%22%20%7Blabel%3D%22..%22%20%22..%2F..%2Fserialization%22%20%5Blabel%3D%22serialization%22%20color%3D%22red%22%20fontcolor%3D%22red%22%20%5D%20%7D%20%7D%0A%20%20%20%20subgraph%20%22cluster_..%22%20%7Blabel%3D%22..%22%20subgraph%20%22cluster_..%2F..%22%20%7Blabel%3D%22..%22%20%22..%2F..%2Ftypes%22%20%5Blabel%3D%22types%22%20color%3D%22red%22%20fontcolor%3D%22red%22%20%5D%20%7D%20%7D%0A%20%20%20%20%22.%2Fmigration_logger%22%20%5Blabel%3D%22migration_logger%22%20color%3D%22red%22%20fontcolor%3D%22red%22%20%5D%0A%20%20%20%20%22boom%22%20%5Blabel%3D%22boom%22%20color%3D%22red%22%20fontcolor%3D%22red%22%20%5D%0A%20%20%20%20%22lodash%22%20%5Blabel%3D%22lodash%22%20color%3D%22red%22%20fontcolor%3D%22red%22%20%5D%0A%20%20%20%20%22lodash.clonedeep%22%20%5Blabel%3D%22lodash.clonedeep%22%20color%3D%22red%22%20fontcolor%3D%22red%22%20%5D%0A%20%20%20%20%22semver%22%20%5Blabel%3D%22semver%22%20color%3D%22red%22%20fontcolor%3D%22red%22%20%5D%0A%20%20%20%20subgraph%20%22cluster_src%22%20%7Blabel%3D%22src%22%20subgraph%20%22cluster_src%2Fcore%22%20%7Blabel%3D%22core%22%20subgraph%20%22cluster_src%2Fcore%2Fserver%22%20%7Blabel%3D%22server%22%20subgraph%20%22cluster_src%2Fcore%2Fserver%2Fsaved_objects%22%20%7Blabel%3D%22saved_objects%22%20subgraph%20%22cluster_src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%22%20%7Blabel%3D%22migrations%22%20subgraph%20%22cluster_src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%2Fcore%22%20%7Blabel%3D%22core%22%20%22src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%2Fcore%2Fdocument_migrator.ts%22%20%5Blabel%3D%22document_migrator.ts%22%20fillcolor%3D%22%23ddfeff%22%20URL%3D%22src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%2Fcore%2Fdocument_migrator.ts%22%5D%20%7D%20%7D%20%7D%20%7D%20%7D%20%7D%0A%0A%20%20%20%20%22src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%2Fcore%2Fdocument_migrator.ts%22%20-%3E%20%22..%2F..%2Fserialization%22%0A%20%20%20%20%22src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%2Fcore%2Fdocument_migrator.ts%22%20-%3E%20%22..%2F..%2Ftypes%22%0A%20%20%20%20%22src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%2Fcore%2Fdocument_migrator.ts%22%20-%3E%20%22.%2Fmigration_logger%22%0A%20%20%20%20%22src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%2Fcore%2Fdocument_migrator.ts%22%20-%3E%20%22boom%22%0A%20%20%20%20%22src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%2Fcore%2Fdocument_migrator.ts%22%20-%3E%20%22lodash%22%0A%20%20%20%20%22src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%2Fcore%2Fdocument_migrator.ts%22%20-%3E%20%22lodash.clonedeep%22%0A%20%20%20%20%22src%2Fcore%2Fserver%2Fsaved_objects%2Fmigrations%2Fcore%2Fdocument_migrator.ts%22%20-%3E%20%22semver%22%0A%7D)

<img width="961" alt="Screen Shot 2019-08-01 at 12 54 16 PM" src="https://user-images.githubusercontent.com/57655/62312187-923cea00-b45b-11e9-99bd-bfbf173e4167.png">

Good luck